File: framework-ts/module/devian-domain-game/features/ability/AbilityUnitMonster.ts
import { UNIT_MONSTER, UNIT_MONSTER_LEVEL } from '../../Generated/Game.g';
import { AbilityUnitBase } from './AbilityUnitBase';

export class AbilityUnitMonster extends AbilityUnitBase {
    private mTable: UNIT_MONSTER | null = null;
    private mLevelTable: UNIT_MONSTER_LEVEL | null = null;

    get unitId(): string { return this.mTable?.unit_id ?? ''; }

    init(table: UNIT_MONSTER, levelTable: UNIT_MONSTER_LEVEL): void {
        this.mTable = table;
        this.mLevelTable = levelTable;
        this.initUnitState(levelTable.unit_level, levelTable.max_hp);
    }

    clone(): AbilityUnitMonster {
        const c = new AbilityUnitMonster();
        c.mTable = this.mTable;
        c.mLevelTable = this.mLevelTable;
        c.copyStatsFrom(this);
        c.copyUnitStateFrom(this);
        return c;
    }
}

File: framework-cs/apps/UnityExample/Assets/Samples/Devian Foundation/0.1.0/GamePackage/Runtime/Ability/AbilityUnitMonster.cs
using Devian.Domain.Game;

namespace Devian
{
    public sealed class AbilityUnitMonster : AbilityUnitBase
    {
        UNIT_MONSTER mTable = null;
        UNIT_MONSTER_LEVEL mLevelTable = null;

        public override string UnitId => mTable?.unit_id ?? string.Empty;

        public void Init(UNIT_MONSTER table, UNIT_MONSTER_LEVEL levelTable)
        {
            mTable = table;
            mLevelTable = levelTable;

            if (levelTable == null)
                return;

            InitUnitState(levelTable.unit_level, levelTable.max_hp);
        }

        public override AbilityBase Clone()
        {
            var c = new AbilityUnitMonster();
            c.mTable = mTable;
            c.mLevelTable = mLevelTable;
            c.CopyStatsFrom(this);
            c.CopyUnitStateFrom(this);
            return c;
        }
    }
}
